cli/src/commands/file/annotate.rs

```diff
@@ -76,8 +79,22 @@ pub(crate) fn cmd_file_annotate(
 ) -> Result<(), CommandError> {
     let workspace_command = command.workspace_helper(ui)?;
     let repo = workspace_command.repo();
+
+    let domain = match &args.domain {
+        None => RevsetExpression::all(),
+        Some(domain) => workspace_command.parse_revset(ui, domain)?.resolve()?,
+    };
     let starting_commit = workspace_command
         .resolve_single_rev(ui, args.revision.as_ref().unwrap_or(&RevisionArg::AT))?;
+
+    if domain
+        .intersection(&RevsetExpression::commit(starting_commit.id().clone()))
+        .evaluate(workspace_command.repo().as_ref())?
+        .is_empty()
+    {
+        return Err(user_error("Starting commit is not contained within domain"));
+    }
```

Comment:
It's probably better to reject non-contiguous range. (See `cmd_diff()` for how we can test that.)

https://github.com/jj-vcs/jj/blob/09d6d4871834cb927d78c4e268bb3508eb673ead/lib/src/annotate.rs#L209